src/core/ucc_progress_queue_mt.c

```diff
@@ -0,0 +1,142 @@
+#include "utils/ucc_atomic.h"
+
+typedef struct ucc_pq_mt {
+    ucc_progress_queue_t super;
+    ucc_spinlock_t       locked_queue_lock;
+    ucc_coll_task_t***   tasks;
+    uint8_t              which_pool;
+    ucc_list_link_t      locked_queue;
+    uint32_t             tasks_countrs[2];
+} ucc_pq_mt_t; // TODO the struct isn't a queue because not maintaining order, maybe another name
+
+#define LINE_SIZE 8 // Could be changed, but in tests performed great
+#define NUM_POOLS 2 // We need extactly two pools
+
+static void ucc_pq_mt_enqueue(ucc_progress_queue_t *pq, ucc_coll_task_t *task) {
+    ucc_pq_mt_t *pq_mt = ucc_derived_of(pq, ucc_pq_mt_t);
+    int i;
+    int which_pool = task->was_progressed ^(pq_mt->which_pool & 1);
+    for (i = 0; i < LINE_SIZE; i++) {
+        if (ucc_atomic_bool_cswap64((uint64_t *) &(pq_mt->tasks[which_pool][i]), 0, (uint64_t) task)) {
+            ucc_atomic_add32(&pq_mt->tasks_countrs[which_pool], 1);
+            return;
+        }
+    }
+    ucc_spin_lock(&pq_mt->locked_queue_lock);
+    ucc_list_add_tail(&pq_mt->locked_queue, &task->list_elem);
+    ucc_spin_unlock(&pq_mt->locked_queue_lock);
+}
+
+static void ucc_pq_mt_dequeue(ucc_pq_mt_t *pq_mt, ucc_coll_task_t **popped_task_ptr, int is_first_call) {
+    int i, curr_which_pool, which_pool;
+    curr_which_pool = pq_mt->which_pool; // extract value in the beginning of the function
+    which_pool = curr_which_pool & 1; // turn from even/odd -> bool
+    ucc_coll_task_t *popped_task = NULL;
+    if (pq_mt->tasks_countrs[which_pool]) {
+        for (i = 0; i < LINE_SIZE; i++) {
+            popped_task = pq_mt->tasks[which_pool][i];
+            if (popped_task) {
+                if (ucc_atomic_bool_cswap64((uint64_t *) &(pq_mt->tasks[which_pool][i]), (uint64_t) popped_task, 0)) {
+                    ucc_atomic_sub32(&pq_mt->tasks_countrs[which_pool], 1);
+                    *popped_task_ptr = popped_task;
+                    popped_task->was_progressed = 1;
+                    return;
+                } else {
+                    i = -1;
+                    // TODO - break statement was here on XCCL, seems like a (perf) mistake - we want the pop to give at least 1 task if exist
+                    // After the MT is fully tested, this comment can be dropped
+                }
+            }
+        }
+    }
+    if (is_first_call) {
+        ucc_atomic_cswap8(&pq_mt->which_pool, curr_which_pool, curr_which_pool + 1);
+        ucc_pq_mt_dequeue(pq_mt, popped_task_ptr, 0);
+        return;
+    }
+    popped_task = NULL;
+    ucc_spin_lock(&pq_mt->locked_queue_lock);
+    if (!ucc_list_is_empty(&pq_mt->locked_queue)) {
+        popped_task = ucc_list_extract_head(&pq_mt->locked_queue, ucc_coll_task_t, list_elem);
+    }
+    ucc_spin_unlock(&pq_mt->locked_queue_lock);
+    if (popped_task) {
+        popped_task->was_progressed = 1;
+    }
+    *popped_task_ptr = popped_task;
+}
+
+ucc_status_t ucc_pq_mt_progress(ucc_progress_queue_t *pq) {
```

Comment:
the problem here is that if we will poll more than 1 task, it will have to be in while loop, and me might end up with 1 thread only progressing tasks and never leave for other tasks.
its possible to just return 0 or 1 maybe.

Reply:
yes, but anyway according to progress queue interface this function returns number of progressed tasks (>=0) or negative number which corresponds to ucc error code